Utilities/Mergers/src/MergerCache.cxx

```diff
@@ -0,0 +1,151 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file MergerCache.cxx
+/// \brief Definition of O2 Merger's Cache
+///
+/// \author Piotr Konopka, piotr.jan.konopka@cern.ch
+
+#include "Mergers/MergerCache.h"
+
+namespace o2
+{
+namespace mergers
+{
+
+//todo: consider storing raw header and payload (not sure if it is deleted by framework)
+MergerCache::MergerCache(bool overwrite) : mCache{}, mOverwrite(overwrite), mTimerPosition(-1), mCachedInputs(0), mUpdatedInputs(0)
+{
+}
+
+MergerCache::~MergerCache()
+{
+  // todo: safely delete collections
+}
+
+void MergerCache::init(const framework::InputRecord& inputs)
+{
+  // sometimes timer can be one of the inputs, but we do not want to store it in the cache. Therefore,
+  // we find its position and omit it when accessing inputs.
+  mTimerPosition = inputs.getPos("timer-publish");
+  mCache = std::vector<CacheEntryQueue>(inputs.size() - (mTimerPosition >= 0));
+
+  mCachedInputs = 0;
+  mUpdatedInputs = 0;
+}
+
+void MergerCache::cacheInputRecord(const framework::InputRecord& inputs) //todo: rename to cacheObjects ?
+{
+  // todo: take care of collections' ownership
+
+  bool shift = false;
+  for (int i = 0; i < inputs.size(); ++i) {
+    // when we iterate besides timer input, we shift the index of cache.
+    if (mTimerPosition == i) {
+      shift = true;
+      continue;
+    }
+
+    auto input = inputs.getByPos(i);
+    if (input.header && input.payload) {
+      auto& queue = mCache[i - shift];
```

Comment:
here I think also it is a bit dangerous to assume the size of "inputs" will be the same as the one you initialized it with - some inputs may come in late (as in - a producer might decide to produce some new data only half way through the fill) - not sure what guarantees the framework gives here (and that may change), but in an async system it is dangerous (and limiting) to make these kind of assumptions.

Reply:
I think we are missing each other a bit on this one. Merger expects one DPL input per producer. Each message for a given input should be a TObject, which can be actually a TCollection of several TObjects, or some custom data structure which can be unpacked to several TObjects. As far as I understood you, the number of TObjects inside one data structure might change, some of them might be missing, which I will take into account.
I remember some discussions about specifying SQL-like requests, but as far as I understood, it would still be 'static' - the whole I/O matching would be done once, during the topology generation. When the topology structure would change (so the number of DPL inputs) I expect that Merger would be reset, so it can reinitialize again.

Reply:
Well, there have been more discussions, not just involving SQL like queries.
The static world-view is the current state of how DPL works, but that might change in the future. The data model (and DPL) allows multiple buffers in an output message, so multiple (T-) objects per message that are completely separate entities, not bound together by a TCollection or anything. Besides - assuming a static configuration is bad practice, especially when it is so easy to code in a more generic, future-proof way. Please don't make hard-coded assumptions on data make-up and contents.